drivers/w1/w1_ds248x.c

```diff
@@ -0,0 +1,445 @@
+
+	return ret > 1 ? 1 : ret;
+}
+
+static int ds248x_write_bit(const struct device *dev, bool bit)
+{
+	int ret;
+
+	ret = ds248x_single_bit(dev, bit ? BIT_SET_msk : BIT_CLR_msk);
+
+	return ret > 0 ? 0 : ret;
+}
+
+static int ds248x_read_byte(const struct device *dev)
+{
+	int ret;
+
+	ret = ds248x_write(dev, CMD_1WRB, NULL);
+	if (ret) {
+		return ret;
+	}
+
+	uint8_t reg;
+
+	do {
+		ret = ds248x_read(dev, REG_NONE, &reg);
+		if (ret) {
+			return ret;
+		}
+	} while (reg & STATUS_1WB_msk);
+
+	ret = ds248x_read(dev, REG_DATA, &reg);
+	if (ret) {
+		return ret;
+	}
+
+	return reg;
+}
+
+static int ds248x_write_byte(const struct device *dev, uint8_t byte)
+{
+	int ret;
+
+	ret = ds248x_write(dev, CMD_1WWB, &byte);
+	if (ret) {
+		return ret;
+	}
+
+	uint8_t reg;
+
+	do {
+		ret = ds248x_read(dev, REG_NONE, &reg);
+		if (ret) {
+			return ret;
+		}
+	} while (reg & STATUS_1WB_msk);
+
+	return 0;
+}
+
+static int ds248x_write_config(const struct device *dev, uint8_t cfg)
+{
+	int ret;
+
+	if (cfg & ~(DEVICE_APU_msk | DEVICE_PDN_msk | DEVICE_SPU_msk | DEVICE_1WS_msk)) {
+		return -EINVAL;
+	}
+
+	uint8_t reg = cfg | ~cfg << 4;
+
+	ret = ds248x_write(dev, CMD_WCFG, &reg);
+	if (ret) {
+		return ret;
+	}
+
+	ret = ds248x_read(dev, REG_NONE, &reg);
+	if (ret) {
+		return ret;
+	}
+
+	return (reg == cfg) ? 0 : -EIO;
+}
+
+static int ds248x_configure(const struct device *dev, enum w1_settings_type type, uint32_t value)
+{
+	switch (type) {
+	case W1_SETTING_SPEED:
+		WRITE_BIT(get_data(dev)->reg_device_config, DEVICE_1WS_pos, value);
+		break;
+	case W1_SETTING_STRONG_PULLUP:
+		WRITE_BIT(get_data(dev)->reg_device_config, DEVICE_SPU_pos, value);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ds248x_write_config(dev, get_data(dev)->reg_device_config);
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int ds248x_pm_control(const struct device *dev, enum pm_device_action action)
+{
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		if (get_config(dev)->slpz_spec.port) {
+			return -ENOTSUP;
+		}
+		if (!device_is_ready(get_config(dev)->slpz_spec.port)) {
+			return -EINVAL;
+		}
+		return gpio_pin_set_dt(&get_config(dev)->slpz_spec, 1);
+	case PM_DEVICE_ACTION_RESUME:
+		if (get_config(dev)->slpz_spec.port) {
+			return -ENOTSUP;
+		}
+		if (!device_is_ready(get_config(dev)->slpz_spec.port)) {
+			return -EINVAL;
+		}
+		return gpio_pin_set_dt(&get_config(dev)->slpz_spec, 0);
+	case PM_DEVICE_ACTION_TURN_OFF:
+	case PM_DEVICE_ACTION_TURN_ON:
+	case PM_DEVICE_ACTION_FORCE_SUSPEND:
+		return -ENOTSUP;
+	};
+
+	return 0;
+}
+#endif /* CONFIG_PM_DEVICE */
+
+static int ds248x_reset_device(const struct device *dev)
+{
+	int ret;
+
+	uint8_t reg;
+
+	ret = ds248x_write(dev, CMD_DRST, NULL);
+	if (ret) {
+		return ret;
+	}
+
+	do {
+		ret = ds248x_read(dev, REG_NONE, &reg);
+		if (ret) {
+			return ret;
+		}
+	} while (!(reg & STATUS_RST_msk));
+
+	return 0;
+}
+
+static int ds248x_init(const struct device *dev)
+{
+	int ret;
+
```

Comment:
would suggest a local var for `get_config(dev)`

Reply:
Only in `init(...)`?

Reply:
Yeah, I was scanning driver from bottom to top.  Personally, I don't think the inline function buys you much w/regards to re-ability vs a local variable for both cfg & data when needed.  (which is what most drivers do).

Reply:
You mean to have a local variable for every function, right? I think this would add a lot of boilerplate code. I don't have (very) strong feelings about this, I'll give it a try and see if the code becomes to noisy.